### PR TITLE
Update TXT.php

### PR DESCRIPTION
**Pull request by MasahikoSada**
To support files encoded by Shift-JIS(Japanese) , GB2312 and GBK(Chinese) .

### CODE REVIEW

**Review by erhuz — CHANGES_REQUESTED** [review]
Would be nice to indent row 70 and remove an indentation from row 69.

**Comment by kesselb** [comment]
Thanks :+1: 

I would say the issue is not the encoding but the font used for txt preview NotoSans-Regular.ttf

On the other hand imagettftext expects a utf8 encoded string as text anyway so a explicit is probably not wrong. Yet https://www.php.net/manual/en/function.mb-convert-encoding should be easier to use here as the from encoding is detected automatically.

File: lib/private/Preview/TXT.php [hunk 1]
```diff
@@ -64,6 +64,12 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
 		if (trim($content) === '') {
 			return null;
 		}
+		
+		$encoding = mb_detect_encoding($content . 'a', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII', true);
+            if ($encoding === '') {
+				$encoding = 'ISO-8859-1';
+				}
```
Comment (kesselb) [inline lib/private/Preview/TXT.php:71]:
```suggestion
```

File: lib/private/Preview/TXT.php [hunk 2]
```diff
@@ -64,6 +64,12 @@ public function getThumbnail(File $file, int $maxX, int $maxY): ?IImage {
...
+		$content = iconv($encoding, 'UTF-8', $content);
```
Comment (kesselb) [inline lib/private/Preview/TXT.php:72]:
```suggestion
		$content = mb_convert_encoding($content, 'UTF-8');
```

**Review by kesselb — CHANGES_REQUESTED** [review]

**Comment by kesselb** [comment]
Thanks @MasahikoSada, can you provide a simple text document or  a string with some characters to test the change?

**Comment by MasahikoSada** [comment]
> Thanks @MasahikoSada, can you provide a simple text document or a string with some characters to test the change?

@kesselb 
Thanks for your reply.

This is a sample file endoced by SHIFT-JIS(ANSI).
https://1drv.ms/t/s!AnT4d94afUwMhYJsm8nD3x9bDGas8w?e=LDAdM7

**Comment by kesselb** [comment]
Hi,

uploaded the test file to my dev instance to generate the preview images. Is one of the previews the expected outcome?

### 1. no patch

txt_master

### 2. version with iconv

txt_iconv

### 3. version with mb_convert_encoding

txt_mb

**Comment by MasahikoSada** [comment]
> Hi,
> 
> uploaded the test file to my dev instance to generate the preview images. Is one of the previews the expected outcome?
> 
> ### 1. no patch
> txt_master
> 
> ### 2. version with iconv
> txt_iconv
> 
> ### 3. version with mb_convert_encoding
> txt_mb

Hi, 
No one is the expected output, but I think that the version 2's encode is proper one.
The only issue of version 2 is just the font.

Can you show me the version  with mb_convert_encoding and 'auto' option.
I think it will be as same as version 2.

**Review by Salvador1313 — APPROVED** [review]

**Comment by kesselb** [comment]
Hey,

Input `���肪�Ƃ��B
����͓��{�ꕶ���̃e�X�g�t�@�C���ł��B`

mb_convert_encoding($content, 'UTF-8', 'auto'); =>  `false`

Blank preview image

mb_convert_encoding($content, 'UTF-8', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII') => `ありがとう。
これは日本語文字のテストファイルです。`

txt_mb_list_of_encodings

**Comment by MasahikoSada** [comment]
> Hey,
> 
> $content = `���肪�Ƃ��B ����͓��{�ꕶ���̃e�X�g�t�@�C���ł��B`
> 
> mb_convert_encoding($content, 'UTF-8', 'auto'); => `false`
> 
> Blank preview image
> 
> mb_convert_encoding($content, 'UTF-8', 'SJIS-win, GB2312, GBK, UTF-8, WINDOWS-1252, ISO-8859-15, ISO-8859-1, ASCII') => `ありがとう。 これは日本語文字のテストファイルです。`
> 

Thanks for your comment.
When 'auto' is set, I think it depends on 'mbstring.language'.
When I set mbstring.language=japanese, the output is expected one.

**Pull request closed without merging**